packages/playwright-core/src/client/mockingProxy.ts

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as network from './network';
+import type * as channels from '@protocol/channels';
+import { ChannelOwner } from './channelOwner';
+import { APIRequestContext } from './fetch';
+import { Events } from './events';
+import { assert } from '../utils';
+import type { Page } from './page';
+
+export class MockingProxy extends ChannelOwner<channels.MockingProxyChannel> {
+  private _pages = new Map<string, Page>();
+
+  constructor(parent: ChannelOwner, type: string, guid: string, initializer: channels.MockingProxyInitializer) {
+    super(parent, type, guid, initializer);
+
+    const requestContext = APIRequestContext.from(initializer.requestContext);
+    this._channel.on('route', async (params: channels.MockingProxyRouteEvent) => {
+      const route = network.Route.from(params.route);
+      route._context = requestContext;
+      this.emit(Events.MockingProxy.Route, route);
```

Comment:
I don't think we can afford events here. The `route` handler must always ensure that the last `_continue()` is called if no handler did run, which is not compatible with events mechanism.